sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,1118 @@
+                # by a factory of ~4.
+                #
+                # Here we are splitting outsamps and insamps into a list
+                # of ranges where both arrays are contiguous. Then we loop
+                # through each sub-range and copy data via slicing.
+                split_idxs = 1 + np.where(
+                    (np.diff(outsamps) > 1) | (np.diff(insamps) > 1))[0]
+                outsplits = np.split(outsamps, split_idxs)
+                insplits = np.split(insamps, split_idxs)
+                for i in range(len(outsplits)):
+                    if len(insplits[i]) == 0:
+                        continue
+
+                    in0, in1 = insplits[i][0], insplits[i][-1] + 1
+                    out0, out1 = outsplits[i][0], outsplits[i][-1] + 1
+
+                    data[:, out0:out1] = iframe['data'].data[:, in0:in1]
+                    biases[:, out0:out1] = iframe['tes_biases'].data[:, in0:in1]
+                    primary[:, out0:out1] = iframe['primary'].data[:, in0:in1]
+                    filled[out0:out1] = 1
+
+                # If there are any remaining samples in the next in_frame, pull it and repeat
+                if np.any((oframe_idxs == oframe_idx) & (iframe_idxs > iframe_idx)):
+                    iframe, interm_frames = next_scan(inframe_iter)
+                    iframe_idx += 1
+                    pbar.update()
+                    continue
+                else:
+                    break
+
+            # Interpolate data where there are gaps
+            if np.all(~filled):
+                self.log.error(
+                    f"No samples properly mapped in oframe frame {oframe_idx}!"
+                    "Cannot properly interpolate for this frame."
+                )
+            elif np.any(~filled):
+                self.log.debug(
+                    f"{np.sum(~filled)} missing samples in out-frame {oframe_idx}"
+                )
+                # Missing samples at the beginning / end of a frame will be
+                # filled with the first / last sample in the frame
+                fill_value = (data[:, filled][:, 0], data[:, filled][:, -1])
+                data[:, ~filled] = interp1d(
+                    ts[filled], data[:, filled], axis=1, assume_sorted=True,
+                    kind='linear', fill_value=fill_value, bounds_error=False
+                )(ts[~filled])
+
+            m = frame_idxs == oframe_idx
+            i0, i1 = np.where(m)[0][[0, -1]]
+
+            oframe = core.G3Frame(core.G3FrameType.Scan)
+
+            if ancil is not None:
+                t0, t1 = ts[0], ts[-1]
+                oframe['ancil'] = ancil.anc_frame_data[oframe_idx]
+                ancil.add_acu_summary_info(oframe, t0, t1)
+
+            oframe['book_id'] = self.book_id 
+            oframe['sample_range'] = core.G3VectorInt([int(i0), int(i1+1)])
+            oframe['flag_smurfgaps'] = core.G3VectorBool(filled)
```

Comment:
Are we flagging interpolated samples (`~filled`) or input samples (`filled`)? (Or did I get these mixed up?)

Reply:
ah yep, good catch